### PR TITLE
MSMAll longitudinal

### PR DESCRIPTION
**Pull request by mmilch01**
Longitudinal adaptation of the MSMAll pipeline and longitudinal batch file for the DeDriftAndResample pipeline.

### CODE REVIEW

File: MSMAll/MSMAllPipeline.sh [hunk 1]
```diff
@@ -216,31 +327,43 @@ log_Msg "fMRIProcSTRING: ${fMRIProcSTRING}"
...
+
+#copy the registration result sphere from template back to timepoints.    
+if (( IsLongitudinal )); then
+    
+    for tp in ${SessionsLong[@]}; do
+        cp $StudyFolder/$TemplateSession/MNINonLinear/Native/$TemplateSession.SphericalDistortion.native.dscalar.nii \
+            $StudyFolder/$SessionLong/MNINonLinear/Native/$SessionLong.SphericalDistortion.native.dscalar.nii
+        for Hemisphere in L R; do
+            SessionLong=$tp.long.$TemplateLong
+            cp $StudyFolder/$TemplateSession/MNINonLinear/Native/$TemplateSession.$Hemisphere.sphere.${OutputRegName}_2_d40_WRN.native.surf.gii $StudyFolder/$SessionLong/MNINonLinear/Native/$SessionLong.$Hemisphere.sphere.${OutputRegName}_2_d40_WRN.native.surf.gii
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:365]: ${OutputRegName}_2_d40_WRN: the _2_d40_WRN of this should not be hard coded.

File: MSMAll/MSMAllPipeline.sh [hunk 2]
```diff
@@ -216,31 +327,43 @@ log_Msg "fMRIProcSTRING: ${fMRIProcSTRING}"
...
+    --msm-all-templates="${MSMAllTemplates}" \
+    --use-ind-mean="${UseIndMean}" \
+    --matlab-run-mode="${MatlabRunMode}"
+
+#copy the registration result sphere from template back to timepoints.    
+if (( IsLongitudinal )); then
+    
+    for tp in ${SessionsLong[@]}; do
+        cp $StudyFolder/$TemplateSession/MNINonLinear/Native/$TemplateSession.SphericalDistortion.native.dscalar.nii \
+            $StudyFolder/$SessionLong/MNINonLinear/Native/$SessionLong.SphericalDistortion.native.dscalar.nii
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:362]: Why is this file being copied?  Is the spherical inflation distortion truly the same across timepoints?  It is possible that it isn't.

File: MSMAll/MSMAllPipeline.sh [hunk 3]
```diff
@@ -138,75 +147,177 @@ log_File_Must_Exist "${TopographyTarget}"
...
 log_Msg "output_proc_string: ${output_proc_string}"
 
-expected_concatenated_output_dir="${StudyFolder}/${Subject}/MNINonLinear/Results/${OutputfMRIName}"
-expected_concatenated_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}${output_proc_string}.dtseries.nii"
-before_vn_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}_novn.dtseries.nii"
+IsLongitudinal=$(opts_StringToBool "$IsLongitudinal")
+
+if (( IsLongitudinal ));  then 
+    
+    if [ -z "$SubjectLong" -o -z "$Sessions" -o -z "TemplateLong" -o -z "${mrfixNamesToUse}" -o -z "$mrfixConcatName" ]; then
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:154]:
```suggestion
    if [[ -z "$SubjectLong" || -z "$Sessions" || -z "TemplateLong" || -z "${mrfixNamesToUse}" || -z "$mrfixConcatName" ]]; then 
```
Same comment as the other PR.

File: MSMAll/MSMAllPipeline.sh [hunk 4]
```diff
@@ -138,75 +147,177 @@ log_File_Must_Exist "${TopographyTarget}"
...
+IsLongitudinal=$(opts_StringToBool "$IsLongitudinal")
+
+if (( IsLongitudinal ));  then 
+    
+    if [ -z "$SubjectLong" -o -z "$Sessions" -o -z "TemplateLong" -o -z "${mrfixNamesToUse}" -o -z "$mrfixConcatName" ]; then 
+        log_Err_Abort "--subject-long, --sessions-long, --template-long are mandatory in longitudinal mode"
+    fi            
+    if [ "$OutputRegName" != "MSMAll_InitalReg" ]; then 
+        log_Err_Abort "Currently, only MSM_All_InitialReg is supported as output reg name in longitudinal mode"
+    fi
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:159]: We need to support arbitrary RegNames.

File: MSMAll/MSMAllPipeline.sh [hunk 5]
```diff
@@ -138,75 +147,177 @@ log_File_Must_Exist "${TopographyTarget}"
...
-before_vn_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}_novn.dtseries.nii"
+IsLongitudinal=$(opts_StringToBool "$IsLongitudinal")
+
+if (( IsLongitudinal ));  then 
+    
+    if [ -z "$SubjectLong" -o -z "$Sessions" -o -z "TemplateLong" -o -z "${mrfixNamesToUse}" -o -z "$mrfixConcatName" ]; then 
+        log_Err_Abort "--subject-long, --sessions-long, --template-long are mandatory in longitudinal mode"
+    fi            
+    if [ "$OutputRegName" != "MSMAll_InitalReg" ]; then 
+        log_Err_Abort "Currently, only MSM_All_InitialReg is supported as output reg name in longitudinal mode"
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:158]:
```suggestion
        log_Err_Abort "Currently, only MSMAll_InitialReg is supported as output reg name in longitudinal mode"
```

**Review by glasserm — CHANGES_REQUESTED** [review]
Please see above.

File: MSMAll/MSMAllPipeline.sh (hunk 4 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:159]: Ok. As long as the Reg name matches cross-sectional, removing this check should be ok.

File: MSMAll/MSMAllPipeline.sh (hunk 1 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:365]: Ok

File: MSMAll/MSMAllPipeline.sh (hunk 2 above)
Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:362]: I guess I am a bit confused here.  Where is this map normally generated from in cross-sectional?  I am not sure that the spherical inflation is exactly the same across the longitudinal timepoints (which is what would be implied by copying this file).

File: MSMAll/MSMAllPipeline.sh (hunk 4 above)
Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:159]: Yes it will match.

File: MSMAll/MSMAllPipeline.sh (hunk 1 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:362]: Ok. I added an option --de-drift-reg-name to put the reg sphere name suffix to.

File: MSMAll/MSMAllPipeline.sh [hunk 6]
```diff
@@ -138,75 +147,174 @@ log_File_Must_Exist "${TopographyTarget}"
...
+    # MNINonLinear/Results/. Multiple configuration files may be used for the same 
+    # subject with different fMRI combinations or longitudinal templates.
+    conf_file="$ResultsTemplateDir/$OutConfig"
+    echo "${TemplateRunsStr/#@/}" > "$conf_file"
+    echo "${TimepointsStr/#@/}" >> "$conf_file"
+    echo "${fMRIRunsStr/#@/}" >> "$conf_file"
+    echo "${ConcatNamesStr/#@/}" >> "$conf_file"
+
+    # average myelin maps from all timepoints.
+    ${average_cmd}
```
Comment (coalsont) [inline MSMAll/MSMAllPipeline.sh:225]: Whitespace safety in command building:

```suggestion
    average_cmd=("${CARET7DIR}/wb_command" -cifti-average \
        "$StudyFolder/$TemplateSession/MNINonLinear/Native/$TemplateSession.$NativeMyelinMap")
        
    for tp in "${SessionsLong[@]}"; do
        SessionLong=$tp.long.$TemplateLong                      #longitudinal session directory name
        average_cmd+=(-cifti "$StudyFolder/$SessionLong/MNINonLinear/Native/$SessionLong.$NativeMyelinMap")
            
        echo "searching $SessionLong for eligible fMRI runs"
        if [ ! -d "$StudyFolder/$SessionLong/MNINonLinear/Results" ]; then 
            log_Err_Abort "ICAFix output does not exist for longitudinal session $SessionLong in $StudyFolder"
        fi    
        ResultsTPLongDir=$StudyFolder/$SessionLong/MNINonLinear/Results
        # iterate over possible fMRI runs and build a list of found runs for this timepoint.
        # Found runs are copied to template directory and all relevant files/dirs are renamed 
        # for the run to be unique within subject.
        # template-based fMRI naming pattern, assuming original run <fMRIRun> label is: <Session>_<fMRIRun>, 
        # where <Session> is longitudinal session label. It is expected to be named <Subject>_<Visit_ID>.
        for fmriName in "${PossibleRuns[@]}"; do
            if [ -d "$ResultsTPLongDir/$fmriName" ]; then            
                TemplateRun=${tp}_${fmriName}
                echo "found $TemplateRun, copying"
                mkdir -p $ResultsTemplateDir/$TemplateRun
                #bulk copy
                cp -r "$ResultsTPLongDir/$fmriName"/* "$ResultsTemplateDir/$TemplateRun/"
                pushd "$ResultsTemplateDir/${TemplateRun}" &> /dev/null                
                for fd in ${fmriName}_* ${fmriName}.*; do
                    [ -e "$fd" ] || continue
                    mv "$fd" "${fd//$fmriName/$TemplateRun}"
                done
                popd &> /dev/null            
                TemplateRunsStr="${TemplateRunsStr}@$TemplateRun"
                TimepointsStr="${TimepointsStr}@$tp"
                fMRIRunsStr="${fMRIRunsStr}@$fmriName"
                ConcatNamesStr="${ConcatNamesStr}@$mrfixConcatName"
            fi
        done
    done
    # Arrays with template fMRI names, matching timepoint labels, original run labels, 
    # and per-timepoint concatenated fMRI names are stored in a configuration file under 
    # MNINonLinear/Results/. Multiple configuration files may be used for the same 
    # subject with different fMRI combinations or longitudinal templates.
    conf_file="$ResultsTemplateDir/$OutConfig"
    echo "${TemplateRunsStr/#@/}" > "$conf_file"
    echo "${TimepointsStr/#@/}" >> "$conf_file"
    echo "${fMRIRunsStr/#@/}" >> "$conf_file"
    echo "${ConcatNamesStr/#@/}" >> "$conf_file"

    # average myelin maps from all timepoints.
    "${average_cmd[@]}"
```

File: MSMAll/MSMAllPipeline.sh [hunk 7]
```diff
@@ -138,75 +147,174 @@ log_File_Must_Exist "${TopographyTarget}"
...
+            if [ -d "$ResultsTPLongDir/$fmriName" ]; then            
+                TemplateRun=${tp}_${fmriName}
+                echo "found $TemplateRun, copying"
+                mkdir -p $ResultsTemplateDir/$TemplateRun
+                #bulk copy
+                cp -r "$ResultsTPLongDir/$fmriName"/* "$ResultsTemplateDir/$TemplateRun/"
+                pushd "$ResultsTemplateDir/${TemplateRun}" &> /dev/null                
+                for fd in ${fmriName}_* ${fmriName}.*; do
+                    [ -e "$fd" ] || continue
+                    mv "$fd" "${fd//$fmriName/$TemplateRun}"
```
Comment (coalsont) [inline MSMAll/MSMAllPipeline.sh:204]: I prefer explicit `if` rather than short-circuit conditionals, as I have to think about (okay, or specifically short-circuits when...the first is true...so that makes this..."if false, then continue"...
```suggestion
                    if [[ -e "$fd" ]]
                    then
                        mv "$fd" "${fd//$fmriName/$TemplateRun}"
                    fi
```

Also, you are looping over glob expressions, the only way this won't exist is if there are no matches for one of them.  Setting `shopt -s nullglob` would mean that you don't need to test for this possibility, because it instead expands to zero strings, instead of the unmodified glob string.

File: MSMAll/MSMAllPipeline.sh [hunk 8]
```diff
@@ -138,75 +147,174 @@ log_File_Must_Exist "${TopographyTarget}"
...
+    done
+    # Arrays with template fMRI names, matching timepoint labels, original run labels, 
+    # and per-timepoint concatenated fMRI names are stored in a configuration file under 
+    # MNINonLinear/Results/. Multiple configuration files may be used for the same 
+    # subject with different fMRI combinations or longitudinal templates.
+    conf_file="$ResultsTemplateDir/$OutConfig"
+    echo "${TemplateRunsStr/#@/}" > "$conf_file"
+    echo "${TimepointsStr/#@/}" >> "$conf_file"
+    echo "${fMRIRunsStr/#@/}" >> "$conf_file"
+    echo "${ConcatNamesStr/#@/}" >> "$conf_file"
```
Comment (coalsont) [inline MSMAll/MSMAllPipeline.sh:222]: I had to look up this syntax to figure out it was "replace at beginning".  We already use the simpler "remove prefix" syntax elsewhere, so I think it is more recognizable.  If it were me, I would store them as arrays and use a subshell with `IFS` to make the string without an extra `@`...
```suggestion
    echo "${TemplateRunsStr#@}" > "$conf_file"
    echo "${TimepointsStr#@}" >> "$conf_file"
    echo "${fMRIRunsStr#@}" >> "$conf_file"
    echo "${ConcatNamesStr#@}" >> "$conf_file"
```
A demo of the `IFS` trick (note, the semicolon and `*` instead of `@` are both important - the subshell stops `IFS` from affecting other commands in the script):
```
foo=(a b c)
(IFS='+'; echo "${foo[*]}")
```

File: MSMAll/scripts/MSMAll.sh [hunk 9]
```diff
@@ -19,10 +19,10 @@ opts_SetScriptDescription "implements MSM-All Registration main script"
 #mandatory
 #general inputs
 opts_AddMandatory '--study-folder' 'StudyFolder' 'path' "folder that contains all subjects" '--path'
-opts_AddMandatory '--subject' 'Subject' '100206' "one subject ID"
+opts_AddMandatory '--session' 'Session' '100206' "one session ID" "--subject"
```
Comment (coalsont) [inline MSMAll/scripts/MSMAll.sh:22]: I don't see any changes to the logic in this script, only a ton of variable replacements.  If this is the case, it would have been simpler to leave the variable name alone:
```suggestion
opts_AddMandatory '--session' 'Subject' '100206' "one session ID" "--subject"
<no changes to the rest of the script>
```
If there are changes in logic, please point them out.

File: MSMAll/scripts/MSMAll.sh [hunk 10]
```diff
@@ -54,6 +54,12 @@ opts_AddOptional '--matlab-run-mode' 'MatlabRunMode' '0, 1, or 2' "defaults to $
 0 = compiled MATLAB
 1 = interpreted MATLAB
 2 = Octave" "$g_matlab_default_mode"
+opts_AddOptional '--is-longitudinal' 'IsLongitudinal' 'TRUE or FALSE' 'Longitudinal mode' "FALSE"
+opts_AddOptional '--longitudinal-sessions' 'SessionsLong' '@ separated list' 'List of @ separated longitudinal sessions' ""
+opts_AddOptional '--longitudinal-template' 'TemplateLong' 'Template label' 'Longitudinal template ID' ""
+opts_AddOptional '--longitudinal-subject' 'SubjectLong' 'Subject ID' 'Subject ID for longtitudinal sessions' ""
```
Comment (coalsont) [inline MSMAll/scripts/MSMAll.sh:60]: None of these appear to get used.

File: MSMAll/MSMAllPipeline.sh [hunk 11]
```diff
@@ -94,6 +95,14 @@ opts_AddOptional '--matlab-run-mode' 'MatlabRunMode' '0, 1, or 2' "defaults to $
 0 = compiled MATLAB
 1 = interpreted MATLAB
 2 = Octave" "$g_matlab_default_mode"
+
+#longitudinal options
+opts_AddOptional '--is-longitudinal' 'IsLongitudinal' 'TRUE or FALSE' "longitudinal mode" "FALSE"
+opts_AddOptional '--subject-long' 'SubjectLong' 'id' "subject ID in longitudinal mode" ""
+opts_AddOptional '--sessions-long' 'Sessions' 'list' "@ separated list of longitudinal timepoints" ""
+opts_AddOptional '--template-long' 'TemplateLong' 'template_id' "longitudinal template ID" ""
+opts_AddOptional '--fmri-out-config-file' 'OutConfig' 'file name' 'Output file with detected fMRI run configuration [fmri_list.txt]' "fmri_list.txt"
```
Comment (coalsont) [inline MSMAll/MSMAllPipeline.sh:104]: Maybe clarify that the user should provide only a bare filename with no path here (since it writes the file relative to MNINonLinear/Results).

Other places in the pipelines, we detect and allow absolute paths even when the default behavior is to assume the path is relative to a specific folder.

File: MSMAll/scripts/SingleSubjectConcat.sh [hunk 12]
```diff
@@ -14,18 +14,23 @@ source "$HCPPIPEDIR/global/scripts/debug.shlib" "$@"         # Debugging functio
...
+opts_AddMandatory '--output-fmri-name' 'OutputfMRIName' 'rfMRI_REST' "name to give to concatenated single session scan"
 opts_AddMandatory '--fmri-proc-string' 'fMRIProcSTRING' 'string' "file name component representing the preprocessing already done, e.g. '_Atlas_hp0_clean'"
 opts_AddMandatory '--output-proc-string' 'OutputProcSTRING' 'string' "the output file name component, e.g. '_vn'"
 #optional inputs
 opts_AddOptional '--start-frame' 'StartFrame' 'integer' "the starting frame to choose from each fMRI run (inclusive), defaults to '1'" '1'
 opts_AddOptional '--end-frame' 'EndFrame' 'integer' "the ending frame to choose from each fMRI run (inclusive), defaults to '' which preserves the ending frame of every fMRI run from --fmri-names-list" ''
+opts_AddOptional '--is-longitudinal' 'IsLongitudinal' 'TRUE or FALSE' "longitudinal mode [FALSE]" "FALSE"
+opts_AddOptional '--subject-long' 'SubjectLong' 'id' "subject ID in longitudinal mode" ""
+opts_AddOptional '--fmri-config-long' 'fMRIConfigLong' 'file_name' "longitudinal runs configuration file [fmri_list.txt]" "fmri_list.txt"
```
Comment (coalsont) [inline MSMAll/scripts/SingleSubjectConcat.sh:31]: Similarly, clarify that the path should not be absolute, or consider whether to add detection and support of absolute paths.

File: MSMAll/MSMAllPipeline.sh (hunk 2 above)
Comment (coalsont) [reply MSMAll/MSMAllPipeline.sh]: Measuring the difference between a timepoint's midthickness and the single longitudinal native sphere can/should produce different results than the longitudinal template midthickness.  On the other hand, since the surface wasn't re-inflated to a sphere for longitudinal surfaces, it can be expected to contain some anatomical changes on top of the actual distortion seen by the spherical inflation algorithm.

Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh]: I don't think we know, but we can ask.

File: Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh
Comment (demsarjure) [inline Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh:1]: @mmilch01 There are no "special" longitudinal flags for DeDriftAndResample, this is pretty much the same as before? Just checking whether I need to change the DeDriftAndResample interface along with the MSMAll one as well.

The pipeline here is that you run the longitudinal MSMAll, followed by "classic" DeDriftAndResample right?

File: MSMAll/scripts/MSMAll.sh (hunk 9 above)
Comment (demsarjure) [reply MSMAll/scripts/MSMAll.sh:22]: Yes, but the "new" variable naming conventions (session vs subject) are aligned with all other scripts. I think it is as good as a time as any to do this refactoring here as well.

**Review by demsarjure — COMMENTED** [review]
@mmilch01 I added some minor comments as well. Please let me know once this is ready for testing. Thanks!

File: MSMAll/MSMAllPipeline.sh (hunk 11 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:104]: Ok.

File: MSMAll/scripts/SingleSubjectConcat.sh (hunk 12 above)
Comment (mmilch01) [reply MSMAll/scripts/SingleSubjectConcat.sh:31]: Ok

File: Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh
Comment (mmilch01) [reply Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh:1]: It is true so far. There's a possibility however that we might need to add longitudinal options, depending on how this thread resolves:
https://github.com/Washington-University/HCPpipelines/pull/334#discussion_r2040738772

File: MSMAll/scripts/MSMAll.sh (hunk 9 above)
Comment (mmilch01) [reply MSMAll/scripts/MSMAll.sh:22]: I see logic in both approaches but since I already did the renaming, will keep it for now

File: MSMAll/MSMAllPipeline.sh [hunk 13]
```diff
@@ -103,7 +103,7 @@ opts_AddOptional '--sessions-long' 'Sessions' 'list' "@ separated list of longit
 opts_AddOptional '--template-long' 'TemplateLong' 'template_id' "longitudinal template ID" ""
 opts_AddOptional '--fmri-out-config-file' 'OutConfig' 'file name' 'Output file with detected fMRI run configuration [fmri_list.txt].
 Specify file name only, without path. The file will be stored under [Session]/MNINonLinear/Results folder.' "fmri_list.txt"
-opts_AddOptional '--dedrift-reg-name' 'DeDriftRegName' 'MSMAll_2_d40_WRN' 'Name part for [Session].[L|R].sphere.[DeDriftRegName].native.surf.gii to be copied to timepoints, to be used in DeDriftAndResample' "MSMAll_2_d40_WRN"
+opts_AddOptional '--dedrift-reg-name' 'DeDriftRegName' 'MSMAll_2_d40_WRN' 'Name part for [Session].[L|R].sphere.[DeDriftRegName].native.surf.gii to be copied to timepoints, to be used in DeDriftAndResample' "MSMAll_InitalReg_2_d40_WRN"
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:106]: This variable should be able to be inferred from existing arguments in this code: _${RegNameStem}_${NumIterations}_d${ICAdim}_${Method} from MSMAll.sh without needing to specify another argument.

File: MSMAll/MSMAllPipeline.sh [hunk 14]
```diff
@@ -230,7 +230,7 @@ if (( IsLongitudinal ));  then
     # variance normalize and concatenate individual runs
     # of all timepoints in template folder. 
     # This script reads the $conf_file.
-    "${HCPPIPEDIR}"/MSMAll/scripts/SingleSubjectConcat.sh \
+    "${HCPPIPEDIR}"/MSMAll/scripts/SingleSubjectConcat.sh
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:233]: Is this really correct?

File: MSMAll/MSMAllPipeline.sh [hunk 15]
```diff
@@ -326,7 +326,7 @@ fMRIProcSTRING+="${output_proc_string}"
 log_Msg "fMRIProcSTRING: ${fMRIProcSTRING}"
 
 # run MSMAll
-"${HCPPIPEDIR}"/MSMAll/scripts/"${ModuleName}" \
+"${HCPPIPEDIR}"/MSMAll/scripts/"${ModuleName}"
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:329]: Is this really correct?

File: MSMAll/MSMAllPipeline.sh [hunk 16]
```diff
@@ -354,12 +354,42 @@ log_Msg "fMRIProcSTRING: ${fMRIProcSTRING}"
...
+
+        ${CARET7DIR}/wb_command -cifti-convert-to-scalar ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dtseries.nii ROW ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dscalar.nii
+        ${CARET7DIR}/wb_command -set-map-name ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dscalar.nii 1 ${SessionLong}_SphericalDistortion
+        ${CARET7DIR}/wb_command -cifti-palette ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dscalar.nii MODE_USER_SCALE ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dscalar.nii -pos-user 0 1 -neg-user 0 -1 -interpolate true -palette-name ROY-BIG-BL -disp-pos true -disp-neg true -disp-zero false
+        rm ${NativeFolderTP}/${SessionLong}.SphericalDistortion.native.dtseries.nii  
+
+        for Mesh in ${HighResMesh} ${LowResMesh} ; do
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:385]: I wonder if we are missing any other initial registration resamplings that currently happen in cross-sectional mode MSMAll at the end?  Alternatively, perhaps we should not resample in longitudinal mode and just take care of that in DeDriftAndResample?

File: Examples/Scripts/MSMAllPipelineBatch-long.sh [hunk 17]
```diff
@@ -124,6 +125,8 @@ mrfixConcatName="fMRI_CONCAT_ALL"
 mrfixNamesToUse="rfMRI_REST1_AP@rfMRI_REST1_PA@rfMRI_REST2_AP@rfMRI_REST2_PA"
 # FIX output concat name for this new MSMAll run of MR FIX
 OutfMRIName="rfMRI_REST"
+# Reg name file used by DeDriftAndResample
+DeDriftRegName="MSMAll_InitalReg_2_d40_WRN"
```
Comment (coalsont) [inline Examples/Scripts/MSMAllPipelineBatch-long.sh:129]: We probably shouldn't put the `InitalReg` typo from YA into the MSMAll example batch (or at least not as the main example).

File: MSMAll/MSMAllPipeline.sh (hunk 13 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:106]: Ok, fixed.

File: MSMAll/MSMAllPipeline.sh (hunk 14 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:233]: Not sure what are you referring to?

File: MSMAll/MSMAllPipeline.sh (hunk 15 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:329]: Ok, fixed \

File: Examples/Scripts/MSMAllPipelineBatch-long.sh (hunk 17 above)
Comment (mmilch01) [reply Examples/Scripts/MSMAllPipelineBatch-long.sh:129]: This typo is also present in both DeDriftAndResample long and cross batches, as well as MSMAllPipeline.sh MSMAllPipelineBatch.sh and MSMAllPipelineBatch-long.sh. I suspect fixing it at one place would trigger a storm of errors on already processed data. So it would probably be better to fix it everywhere. Or, leave it as is? @glasserm  @coalsont @mharms let me know what you think would be the least painful.

File: MSMAll/MSMAllPipeline.sh (hunk 16 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:387]: Conceptually these two approaches are equivalent, we'll need to isolate all resampling code that needs to be run either in DeDrift.. or in MSMAllPipeline.sh under if (( IsLongitudinal )). DeDrift.. does more than just resampling so I guess it would depend on predominant workflow preference. Doing it in MSMAllPipeline has advantage that it's a clean cut for longitudinal mode, equivalent to what is now implemented cross-sectionally.

File: MSMAll/MSMAllPipeline.sh (hunk 13 above)
Comment (coalsont) [reply MSMAll/MSMAllPipeline.sh:106]: You need to remove the option flag, too.

File: Examples/Scripts/MSMAllPipelineBatch-long.sh (hunk 17 above)
Comment (coalsont) [reply Examples/Scripts/MSMAllPipelineBatch-long.sh]: Lifespan changed it to `Initial`.  Examples that use YA should stay `Inital`, but it would be nice to have lifespan examples as the primaries, and have all the typo examples commented out...but, this may be out of scope.  For this PR, though, it would make sense for the example to be lifespan, not YA (because YA wasn't really designed to collect longitudinal), which also means the example wouldn't need the typo.

File: MSMAll/MSMAllPipeline.sh (hunk 16 above)
Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:387]: I think we can not resample initials in the longitudinal timepoints in MSMAll and just resample finals in DeDriftAndResample.

File: Examples/Scripts/MSMAllPipelineBatch-long.sh (hunk 17 above)
Comment (glasserm) [reply Examples/Scripts/MSMAllPipelineBatch-long.sh]: I agree.

Comment (mharms) [reply Examples/Scripts/MSMAllPipelineBatch-long.sh]: Also, this is just an example script.  There is no reason that it needs to contain a typo, even if that typo happens to have made it into the actual processing of the HCP-YA data.

Comment (glasserm) [reply Examples/Scripts/MSMAllPipelineBatch-long.sh]: I agree.

File: MSMAll/MSMAllPipeline.sh (hunk 13 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh]: Thought I did. Done.

File: MSMAll/MSMAllPipeline.sh (hunk 16 above)
Comment (mmilch01) [reply MSMAll/MSMAllPipeline.sh:386]: > I wonder if we are missing any other initial registration resamplings that currently happen in cross-sectional mode MSMAll at the end? Alternatively, perhaps we should not resample in longitudinal mode and just take care of that in DeDriftAndResample?

Ok, to clarify, should we move all resampling from MSMAll to DeDrift.., or just do it in longitudinal case?

Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:386]: Just longitudinal.

File: Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh
Comment (mmilch01) [reply Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh:1]: > @mmilch01 There are no "special" longitudinal flags for DeDriftAndResample, this is pretty much the same as before? Just checking whether I need to change the DeDriftAndResample interface along with the MSMAll one as well.
> 
> The pipeline here is that you run the longitudinal MSMAll, followed by "classic" DeDriftAndResample right?

Another update: please hold with your tests for DeDriftAndResample. Based on discussion below, I'll create a separate longitudinal branch for DeDriftAndResample, there will be some modifications there for longitudinal mode.

Comment (glasserm) [reply Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh:1]: No modifications are needed.

File: MSMAll/MSMAllPipeline.sh (hunk 16 above)
Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:386]: All you need to do is run DeDriftAndResample after creating the unresampled spherical distortion maps.

Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:386]: I think we need to discuss before you do anything.

Comment (glasserm) [reply MSMAll/MSMAllPipeline.sh:386]: This was discussed over the phone and a plan will be implemented and tested.

File: Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh
Comment (mmilch01) [reply Examples/Scripts/DeDriftAndResamplePipelineBatch-long.sh:1]: Yes, sorry for the confusion, the original plan to not modify DeDriftAndResample stays.

**Review by glasserm — APPROVED** [review]
@demsarjure can you run your tests now?

**Review by glasserm — CHANGES_REQUESTED** [review]
Jure found bugs in Slack.

File: MSMAll/MSMAllPipeline.sh [hunk 18]
```diff
@@ -138,75 +147,178 @@ log_File_Must_Exist "${TopographyTarget}"
 output_proc_string="_vn" #To VN only to indicate that we did not revert the bias field before computing VN
 log_Msg "output_proc_string: ${output_proc_string}"
 
-expected_concatenated_output_dir="${StudyFolder}/${Subject}/MNINonLinear/Results/${OutputfMRIName}"
-expected_concatenated_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}${output_proc_string}.dtseries.nii"
-before_vn_output_file="${expected_concatenated_output_dir}/${OutputfMRIName}${fMRIProcSTRING}_novn.dtseries.nii"
+IsLongitudinal=$(opts_StringToBool "$IsLongitudinal")
+NumIterations=$(echo "${Iterations}" | wc -w)
```
Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:151]: `NumIterations=$(echo "${IterationModes}" | sed s/_/ /g | wc -w)`

Comment (glasserm) [inline MSMAll/MSMAllPipeline.sh:151]:
```suggestion
NumIterations=$(echo "${IterationModes}" | sed s/_/ /g | wc -w)
```

**Review by glasserm — APPROVED** [review]
